### PR TITLE
Enable NPC Movement

### PR DESCRIPTION
**Pull request by AricStewart**
Address  Trello#29 https://trello.com/c/hQnlj8Jj/29-npc-movement

I expect this large of a merge will require some cleaning up and comments.

But I need to get the process started.

### CODE REVIEW

**Comment by yelworc** [comment]
Okay, here we go :)

Correct me if I'm wrong, but it seems to me that we could generally instantiate the `movement` property in the Item constructor. Following that, I would suggest a bit of restructuring, moving the logic that is currently in `ItemApi` into `ItemMovement`, creating two new functions there:

* `startMoving(dest, options)` _(for apiFindPath and apiMoveToXY)_
where `dest` is an `{x, y}` object, and `options` may contain a combination of other parameters for the specific movement call (flags, callback, ...)

* `startFlying(dest, options)` _(for apiStartFlyingInTheArea, apiStartFlyingInTheAreaX, apiStartFlyingTo, apiKickTheChicken)_
where `dest` is an object (either `{x, y}` or `{x, y, w, h}`), and `options` may contain other movement parameters (stopAtEnd, changeState, speed, kickVX, kickVY, callback, ...)

These functions should create paths and initialize/start the movement process according to their parameters. They would also replace some of the existing ItemMovement functions (`kickTheChicken`, `moveToXY`).

`apiStopMoving` could probably just call `ItemMovement.stopMove` directly, and we can get rid of `ItemMovement.stopMoving` I think? (And I would rename the remaining function to `stopMoving`, for consistency)

`setItemTimeout` should be replaced with `GameObject.setGsTimer`, with the `internal` option (and the `clearTimeout` call in `stopMove` with `GameObject.cancelGsTimer`, respectively).

Ok... haven't really looked into the core mechanisms yet (`moveStep` and the `moveFlying`/`Walking`/`Direct` functions), but I guess this is enough for now anyway.
Sorry, this sounds like a lot of changes, but generally I'm happy about the direction this is going!
Also, don't worry too much about the tests while we're making these structural changes.

Oh, and please use the linter (`npm -s run lint`) :)

**Comment by AricStewart** [comment]
Oh thanks for the linter command. It is much more strict than my jshint configuration. I will have to see if i can do a project jshint configuration to matches this.

Looking at the other structuring things.  Do we want every item to instantiate the movement property even if they are not an npc that would be moving? I am not sure what the overhead on javascript would be.  I guess if it all just prototype functions the overhead would not be too bad. 

I am liking the idea of reducing the ItemMovement entry points and having the api* functions just build structures and pass them in. 

I am thinking actaully just a single startMoving function and even the transport {flying, walking etc} is passed in.

**Comment by yelworc** [comment]
If we can find a not-too-hacky way to only give the `movement` property to items that actually move, that would be good (although I don't _think_ the overhead would be terrible). Not sure if there is a good way though - I had something like this in mind first:
`this.parent_classes.indexOf('npc_walkable') !== -1`
But that does not include e.g. salmon.

And yes, a single `startMoving` entry point sounds good too (probably dispatching calls to more specific movement start functions internally, if it becomes too complicated).

**Comment by AricStewart** [comment]
I get the impression that you feel the current way of creating the movement property on first call of a movement functions to be hacky then. I will brainstorm over it and see if i come up with anything reasonable.

**Comment by yelworc** [comment]
Well now that you put it this way, it does make a lot of sense — when a movement function is called on it, it's probably an item that should move (if it quacks...).
Ok, how about this then: two very minimal functions `Item.startMoving` and `Item.stopMoving`, which take care of creating the `ItemMovement` property when necessary (just in the "start" function I guess), and then forward any arguments to their respective counterparts in `ItemMovement`?

**Comment by AricStewart** [comment]
Ok, cleaned up, restructured and re-pushed.

Ready for another review.

**Comment by AricStewart** [comment]
Ok re-pushed with changes and ready for re-review

File: src/model/ItemMovement.js [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+	if (this.callback) {
+		if (this.flags && 'callBackParams' in this.flags) {
+			fs = this.callback.call(this.item, this.flags.callBackParams);
+		}
+		else {
+			fs = this.callback.call(this.item, {status: status});
+		}
+	}
+	return fs;
+};
```
Comment (yelworc) [inline src/model/ItemMovement.js]: What exactly is `fs`?

File: src/model/ItemMovement.js [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+	return true;
+};
+
+/*
+ * Stop item movement.
+ */
+ItemMovement.prototype.stopMovement = function stopMovement() {
+	log.debug('%s.apiStopMoving()', this);
+	this.stopMove(MOVE_CB_STATUS.STOP);
+	return true;
```
Comment (yelworc) [inline src/model/ItemMovement.js]: That return is unnecessary AFAICS (at least `Item.prototype.stopMoving` does not use it)

File: src/model/ItemMovement.js [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+	return nextStep;
+};
+
+/**
+ * Moves the item via the transportation dictated in the next path point
+ *
+ * @param {object} nextPath the next pathing target
+ * @returns {object} the next step in the path
+ */
+ItemMovement.prototype.transport = function transport(nextPath) {
```
Comment (yelworc) [inline src/model/ItemMovement.js]: I think this function isn't used anywhere

File: src/model/ItemMovement.js [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+};
+
+
+/**
+ * Start chicken like movement.
+ *
+ * @param {number} vx velocity of the x movment
+ * @param {number} vy velocity of the y movment
+ */
+ItemMovement.prototype.kickTheChicken = function kickTheChicken(vx, vy) {
```
Comment (yelworc) [inline src/model/ItemMovement.js]: I think this function isn't used anywhere

File: src/model/ItemMovement.js [hunk 5]
```diff
@@ -0,0 +1,675 @@
...
+	this.item.setGsTimer({fname: 'movementTimer', delay: 333, interval: true,
+		internal: true});
+	return true;
+};
+
+/*
+ * Stop item movement.
+ */
+ItemMovement.prototype.stopMovement = function stopMovement() {
+	log.debug('%s.apiStopMoving()', this);
```
Comment (yelworc) [inline src/model/ItemMovement.js]: copypaste :)

**Comment by yelworc** [comment]
Apart from the minor detail things above:

In `ItemMovement.prototype.kickCallback`, I would replace the `apiStartFlyingTo` calls with the respective `this.startMove` calls. Seems cleaner to me, and that way I think we could also get rid of `Item.prototype.movementCallback`.

Regarding the `callBackParams` argument in `apiMoveToXY`: Although it's not used in GSJS (as far as I can see), I guess `callBackParams` is supposed to allow an arbitrary number of additional arguments. See e.g. the timer API functions in `GameObjectApi` on how to parse and document those. The call in `ItemMovement.prototype.stopMove` would have to be adjusted, too (something like `this.callback.apply(this.item, this.flags.callBackParams)`).
But since this is only ever used in GSJS code with a single numeric argument, we might as well redefine the API a bit here, and just rename the parameter to `callBackParam` (singular); I'm okay with both options.

Another small idea: Maybe define constants (`LEFT`/`RIGHT`/`ABOVE`/`BELOW`) for 1/-1 in `ItemMovement`? Only if you agree it increases readability a little, otherwise ignore this.

File: src/model/ItemMovement.js (hunk 1 above)
Comment (AricStewart) [inline src/model/ItemMovement.js]: It was fullStatus, but it is if for queueChanges we set a full report or not. I can rename it.

File: src/model/ItemMovement.js (hunk 3 above)
Comment (AricStewart) [inline src/model/ItemMovement.js]: Line 462

var nextStep = this.transport(this.path[0]);

File: src/model/ItemMovement.js (hunk 5 above)
Comment (AricStewart) [inline src/model/ItemMovement.js]: Not sure the issue.  Interval vs Internal?

**Comment by AricStewart** [comment]
Thank you for the review,  code reworked (unless addressed in comment above) and resent!

thanks!

**Comment by yelworc** [comment]
Thanks, looking good so far - I'm currently reading through it again (sort of a last pre-merge pass), and adjusting style and docs here and there as I go along. If it turns out to change more than expected, I'll send it back to you for cross-checking.

**Comment by AricStewart** [comment]
Great thanks for all the help!

**Pull request merged**